orangecontrib/text/preprocess.py

```diff
@@ -1,94 +1,219 @@
+        if transformation is not None and not isinstance(transformation, Stemmatizer):
+            raise ValueError("Type '{}' not supported.".format(transformation))
+        self.transformation = transformation
 
     def __call__(self, data):
+        if isinstance(data, Corpus):
+            data = data.documents
+
         if isinstance(data, str):
-            vec = self.cv.fit([data])
-            return vec.get_feature_names()
-        if isinstance(data, list):
-            vec = self.cv.fit(data)
-            features = vec.get_feature_names()
-            docs = [[] for _ in range(len(data))]
-            for (line, column), count in vec.transform(data).todok().items():
-                docs[line].extend([features[column]] * count)
-            return docs
+            pp_corpus = self._preprocess_document(data)
+            return pp_corpus
+        elif isinstance(data, list):
+            if self.use_df_sw:  # DF parameters were specified.
+                self._get_corpus_specific_stopwords(data)
+            pp_corpus = [self._preprocess_document(doc) for doc in data]
+            return pp_corpus
         else:
             raise ValueError("Type '{}' not supported.".format(type(data)))
 
+    def _preprocess_document(self, data):
+        # Tokenize.
+        data = self._tokenize(data)
+        # Lowercase.
+        if self.lowercase:
+            data = [token.lower() for token in data]
+        # Remove stop words.
+        if self.stop_words is not None:
+            data = self._remove_stop_words(data, self.stop_words)
+        if self.df_stop_words is not None:
+            data = self._remove_stop_words(data, self.df_stop_words)
+        # Transform.
+        if self.transformation is not None:
+            data = self._transform(data)
+        return data
+
+    def _tokenize(self, data):
+        """
+        Splits the text of the input into tokens by whitespaces and punctuation.
+        :param data: The input holding the text to be tokenized.
+        :type data: str
+        :return: list
+        """
+        if isinstance(data, str):
+            return self.tokenizer(data)
+        else:
+            raise ValueError("Type {} not supported.".format(type(data)))
+
+    def _transform(self, data):
+        """
+        Performs on the input, whatever transformation was specified at class instance creation.
+        :param data: The input holding the text to be transformed. Can be provided
+            in form of a string or list of string tokens.
+        :type data: string, list
+        :return: string, list
+        """
+        return self.transformation(data)
+
+    @staticmethod
+    def _remove_stop_words(data, stop_words):
+        """
+        Removes the stop words from the corpus. Removes either the stop words specified
+        at class instance creation, or stop words calculated from the min/max_df parameters.
+        """
+        return [word for word in data if word not in stop_words]
```

Comment:
Please add here, above the return, a line:
```python
assert isinstance(stop_words, set)
```
No point in making it much slower that it needs to be. :)